netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+    private Queue<ChannelWrite> bufferedWrites = new ArrayDeque<ChannelWrite>();
+    private boolean flushed;
+
+    /**
+     * @param handlers the ChannelHandlers are added to the pipeline on channelRegistered and
+     *                 before this handler.
+     */
+    BufferWritesHandler(ChannelHandler... handlers) {
+      this.handlers = handlers;
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+      /**
+       * This check is necessary as a channel may be registered with different eventloops during
+       * its lifetime and we only want to set up the pipeline once.
+       * We can't set up the pipeline by specifying a ChannelInitializer in the client bootstrap,
+       * because the initializer is not executed in the calling thread but on channelRegistered
+       * in the eventloop thread. So there is a race condition between threads immediately writing
+       * to the channel (from outside the eventloop) and the pipeline being set up i.e. a
+       * write task (writing to the HeadContext) may be created before the pipeline is set up.
+       */
+      if (handlers != null) {
+        ctx.pipeline().addFirst(handlers);
+        handlers = null;
       }
-    };
+      super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+      failBufferedAndClose(ctx);
+      super.channelInactive(ctx);
+    }
 
-    return new Negotiation() {
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+        throws Exception {
+      /**
+       * This check is necessary as there is a race condition between threads writing to the
+       * channel and the handler removing itself from the pipeline (once all buffered messages
+       * are written) i.e. channel.write(msg) may add a write task to the eventloop that invokes
+       * write on the ChannelHandlerContext for this handler, but before the write task is executed
+       * this handler may remove itself from the pipeline. For details have a look at the
+       * implementation of AbstractChannelHandlerContext#write(Object msg, ChannelPromise promise).
+       */
+      if (bufferedWrites == null) {
+        super.write(ctx, msg, promise);
+      } else {
+        bufferedWrites.add(new ChannelWrite(msg, promise));
       }
+    }
 
-      @Override
-      public ListenableFuture<Void> completeFuture() {
-        return completionHandler.getUpgradeFuture();
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+      /**
+       * Don't propagate flush while there are still messages in the buffer.
+       * This is required to make buffering with the SslHandler work, as otherwise
+       * TLS handshakes "randomly" timeout. The SslHandler doesn't like flushes while the
```

Comment:
@nmittler well it's SSL specific in the sense that it breaks the `SslHandler`, because that guy issues a write for every flush which can make the handshake fail [1]. So that's why this comment is there, to make devs looking at this in the future that flushing breaks the `SslHandler`. I guess i should formulate it differently.

In order to be correct we need the `if (bufferedWrites == null)` check before we flush, for the same reason we need the check in `write`. Now we can add a else clause that sets `flushRequested`. I didn't find it necessary because we always need to flush in grpc anyway. 

[1] https://github.com/netty/netty/blob/master/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L428